subsys/bluetooth/controller/bluetooth_controller.cmake

```diff
@@ -0,0 +1,60 @@
+set_target_properties(subsys_bluetooth PROPERTIES EXCLUDE_FROM_ALL False)
+
+zephyr_populate_source_list(
+  OUTPUT PRIVATE_SOURCES
+  util/mem.c
+  util/memq.c
+  util/mayfly.c
+  util/util.c
+  ticker/ticker.c
+  ll_sw/ll_addr.c
+  ll_sw/ll_tx_pwr.c
+  crypto/crypto.c
+  hci/hci_driver.c
+  hci/hci.c
+  IFDEF:${CONFIG_BT_CTLR_FILTER} ll_sw/ll_filter.c
+  IFDEF:${CONFIG_SOC_FAMILY_NRF} hal/nrf5/cntr.c
+                                 hal/nrf5/ecb.c
+                                 hal/nrf5/radio/radio.c
+                                 hal/nrf5/mayfly.c
+                                 hal/nrf5/ticker.c
+)
+
+if(CONFIG_BT_LL_SW)
+  zephyr_populate_source_list(
+    OUTPUT PRIVATE_SOURCES
+    APPEND
+    ll_sw/ctrl.c
+    ll_sw/ll.c
+    IFDEF:${CONFIG_BT_BROADCASTER} ll_sw/ll_adv.c
+    IFDEF:${CONFIG_BT_OBSERVER}    ll_sw/ll_scan.c
+    IFDEF:${CONFIG_BT_CENTRAL}     ll_sw/ll_master.c
+    IFDEF:${CONFIG_BT_CTLR_DTM}    ll_sw/ll_test.c
+  )
+endif()
+
+target_sources(subsys_bluetooth PRIVATE   ${PRIVATE_SOURCES})
+
+#
+# The lines below can be updated in similar way, if principle is approved.
+#
+target_include_directories(subsys_bluetooth PRIVATE
+  ${CMAKE_CURRENT_LIST_DIR}/.
+  ${CMAKE_CURRENT_LIST_DIR}/util
+  ${CMAKE_CURRENT_LIST_DIR}/hal
+  ${CMAKE_CURRENT_LIST_DIR}/ticker
+  ${CMAKE_CURRENT_LIST_DIR}/include
+)
+
+# This way of overriding compile flags for specific sources is safer than the link libraries.
+# That flags are always added in the end by CMake, thus those flags will have precedence over any
+# library flags. This automatically takes care of what is discussed in: #5097
+# Also it prevents any unintended propagation of flags as described in: #<>
+if(CONFIG_BT_CTLR_FAST_ENC)
+  SET_SOURCE_FILES_PROPERTIES( ${PRIVATE_SOURCES} PROPERTIES COMPILE_FLAGS -Ofast)
+endif()
+
+# We must tell the linker not to remove the _hci_driver_init function during linking.
+# Notice: This is done in the same file where we have the knowledge that hci_driver.c is added to
+#         the library target.
```

Comment:
The link command must be kept as clean as possible.

Perhaps we could achieve this by modifying SYS_INIT and/or the linker scripts to force symbols into the build.

Reply:
You need someone OUTSIDE the library to use the symbol.
Just placing an internal reference won't help you.

Placing it as a link dependency actually makes it very clear, that this is needed when using this module.
And the flag is activated based on the KConfig setting, which includes the file requiring the flag.
In this case: hci_driver.c and -u_hci_driver_init are only added to the library when CONFIG_BT_CTRL is enable in KConfig.
This makes things that belongs together also being placed together.

Reply:
> Placing it as a link dependency actually makes it very clear, that this is needed when using this module.

I agree that it reads cleanly that this flag is needed when using this module. What I want is for this flag to not be needed at all.

Inuitively, users should as often as possible be able to work with a C-based interface, like the SYS_INIT macro. And then the build will behave as they expect it to without them having to also keep in sync metadata in some other file, like this build script.

It is redundant to invoke SYS_INIT and also to say that this symbol is undefined. This redundancy leads to user errors when these two pieces of information get out-of-sync. This redundancy also hurts the learning curve as the user must learn that this is necessary through documentation.

Reply:
It seems we will not have to fake an unresolved reference for all drivers.

E.g. bt_hci_driver can be resolved like this:

https://github.com/SebastianBoe/zephyr/commit/d8d6fd7835a9715fceec7455e397e2894a912d97

This has the added benefit of cutting down a bit on runtime overhead and flash size.

Perhaps if it is rare that the user needs to insert this reference this is acceptable ...

I can't think of any simple mechanisms that would avoid this metadata (although several complex ones could ...).

@jhedberg has confirmed that it is OK to drop (it doesn't actually exist) runtime-support for multiple driver implementations. So the above patch at least should be safe ...

Reply:
So your suggestion is to remove of a struct containing the function, with directly called functions.

It's good that @jhedberg has confirmed that this is ok in this particular case, but will this be a solution that can work everywhere ?

The reason I ask is because having to write code in different way seems like limiting the features of C in Zephyr.

Currently, I have seen the problem with the _hci_driver_init in relation to the SYS_INIT macro.
I currently don't know how many places the SYS_INIT is used in a way that causes same problem as with the _hci_driver_init, but SYS_INIT is used 191 times i C code.
And some of those might not be able to change this way.

However, it will be safe to add -u<function> to those that benefits from it, such as _hci_driver_init.

Get me right on this: Simplifying code when it can be done, while at the same time remove this problem is good.
But we should be careful not to enforce people to write drivers in a specific way or limit the features of C.

Reply:
@SebastianBoe Please pay attention to the comment provided by @andrewboie 

> Have you checked the effect on various footprint sizes when modifying the archive settings? Are there any differences? I mention this because we use these settings to save space in the kernel. For example, if the application does not use any of the workqueue APIs, then the code in kernel/system_work_q.c, instead of pulling the rest of the workqueue code in, gets dropped too and the k_sys_work_q_init function never gets run at boot.

SYS_INIT is not always the same as the symbol should be considered undefined by linker.
In some cases (kernel), the code should be dropped

Reply:
The commit: ffa3a2b introduces a property you can set on the target.

This will search the target for the devconfig.sysinit section after creation and any init symbols will be populated into the linker.

This removes the need for knowing/specifying the exact symbol names, thus:
`target_link_libraries(<target> INTERFACE -u<symbol> -u<...> ...)`
 is no longer needed.

Instead, a property is set on the target when it is known to contain initialization symbols:
`zephyr_set_property(TARGET <target> SYSINIT)`
eg.
`zephyr_set_property(TARGET subsys_bluetooth SYSINIT)`